04-Protocol-Security.md

```diff
@@ -0,0 +1,407 @@
+   2. creates two new CipherState objects `c1` and `c2`
+   3. calls `c1.InitializeKey(temp_k1)` and `c2.InitializeKey(temp_k2)`
+   4. returns the pair `(c1, c2)`
+
+```
+Message format of NX-handshake part 2 
++-------------------------+--------------------------------------------------------------------------------------------+
+| PUBKEY                  |  Responder's plaintext ephemeral public key                                                |
++-------------------------+--------------------------------------------------------------------------------------------+
+| PUBKEY                  |  Responder's encrypted static public key                                                   |
++-------------------------+--------------------------------------------------------------------------------------------+
+| MAC                     |  Message authentication code for responder's static public key                             |
++-------------------------+--------------------------------------------------------------------------------------------+
+| SIGNATURE_NOISE_MESSAGE |  Signed message containing Responder's static key. Signature is issued by authority that   |
+|                         |  is generally known to operate the server acting as the noise responder                    |
++-------------------------+--------------------------------------------------------------------------------------------+
+| MAC                     |  Message authentication code for SIGNATURE_NOISE_MESSAGE                                   |
++-------------------------+--------------------------------------------------------------------------------------------+
+
+Message length: 170 bytes
+```
+
+#### 4.5.2.2 Initiator
+1. receives NX-handshake part 2 message
+2. interprets first 32 bytes as `re.public_key`
+3. calls `MixHash(re.public_key)`
+4. calls `MixKey(ECDH(e, re))`
+5. decrypts next 48 bytes with `DecryptAndHash()` and stores the results as `rs.public_key` which is **server's static public key** (note that 32 bytes is the public key and 16 bytes is MAC)
+6. calls `MixKey(ECDH(e, rs)`
+7. decrypts next 90 bytes with `DecryptAndHash()` and deserialize plaintext into `SIGNATURE_NOISE_MESSAGE` (74 bytes data + 16 bytes MAC)
+9. return pair of CipherState objects, the first for encrypting transport messages from initiator to responder, and the second for messages in the other direction:
+   1. sets `temp_k1, temp_k2 = HKDF(ck, zerolen, 2)`
+   2. creates two new CipherState objects `c1` and `c2`
+   3. calls `c1.InitializeKey(temp_k1)` and `c2.InitializeKey(temp_k2)`
+   4. returns the pair `(c1, c2)`
+
+### 4.5.3 Server authentication
+Identity of the server is confirmed by initiator by verifying the signature in `CERTIFICATE`.
+Certificate is constructed from the `SIGNATURE_NOISE_MESSAGE`, authority public key that is generally known (for example from pool's website) and **server's static public key** that has been received during `NX` handshake.
```

Comment:
Maybe you can specify that is the authority that signs the certificate, so the certificate's 
signatures has to be checked against the pubkey of trusted authority.